src/processing/evm.rs

```diff
@@ -0,0 +1,198 @@
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use alloy::hex::FromHex;
+use alloy::primitives::FixedBytes;
+use arc_swap::Guard;
+use bigdecimal::ToPrimitive;
+use num_bigint::BigInt;
+use starknet::core::types::{BlockId, BlockTag, Felt, FunctionCall};
+use starknet::macros::selector;
+use starknet::providers::jsonrpc::HttpTransport;
+use starknet::providers::{JsonRpcClient, Provider};
+
+use crate::config::{Config, EvmConfig};
+use crate::constants::EVM_TIME_SINCE_LAST_FEED_UPDATE;
+use crate::{config::get_config, error::MonitoringError};
+
+#[derive(Clone)]
+enum FeedType {
+    Unique(UniqueVariant),
+    Twap(TwapVariant),
+}
+
+impl FeedType {
+    async fn update_data(&self, chain: &EvmConfig, feed_id: FeedId) -> Result<(), MonitoringError> {
+        match self {
+            FeedType::Unique(unique_variant) => unique_variant.update_data(chain, feed_id).await,
+            FeedType::Twap(twap_variant) => twap_variant.update_data(chain, feed_id).await,
+        }
+    }
+}
+
+#[derive(Clone)]
+enum UniqueVariant {
+    SpotMedian,
+    PerpMedian,
+}
+
+#[derive(Clone)]
+enum TwapVariant {
+    SpotTwap = 0,
+}
+
+impl TryFrom<Felt> for FeedType {
+    type Error = MonitoringError;
+
+    fn try_from(feed_id: Felt) -> Result<Self, Self::Error> {
+        let feed_type = ((feed_id.to_bigint() & BigInt::from(65280)) / BigInt::from(256))
+            .to_u64()
+            .ok_or(MonitoringError::Evm("invalid main type".to_string()))?;
+        let variant =
+            (feed_id.to_bigint() & BigInt::from(255))
+                .to_u64()
+                .ok_or(MonitoringError::Evm(format!(
+                    "invalid variant for main type {}",
+                    feed_type
+                )))?;
+        match feed_type {
+            0 => match variant {
+                0 => Ok(FeedType::Unique(UniqueVariant::SpotMedian)),
+                1 => Ok(FeedType::Unique(UniqueVariant::PerpMedian)),
+                _ => unreachable!(),
+            },
+            1 => match variant {
+                0 => Ok(FeedType::Twap(TwapVariant::SpotTwap)),
+                _ => unreachable!(),
+            },
+            _ => unreachable!(),
+        }
+    }
+}
+
+pub trait Variant: Clone + Send + Sync {
+    async fn update_data(&self, chain: &EvmConfig, feed_id: FeedId) -> Result<(), MonitoringError>;
```

Comment:
why `update_data`? confusing, looks like we are updating the data. should be either
* get_latest_data
* get_latest_updated_data
smth like that

Reply:
done